tools/pipeline-witness/Azure.Sdk.Tools.PipelineWitness/Configuration/PostConfigureKeyVaultSettings.cs

```diff
@@ -0,0 +1,60 @@
+﻿using System;
+using System.Linq;
+using System.Text.RegularExpressions;
+
+using Microsoft.Extensions.Logging;
+using Microsoft.Extensions.Options;
+
+namespace Azure.Sdk.Tools.PipelineWitness.Configuration
+{
+    public class PostConfigureKeyVaultSettings<T> : IPostConfigureOptions<T> where T : class
+    {
+        private static readonly Regex secretRegex = new Regex(@"(?<vault>https://.*?\.vault\.azure\.net)/secrets/(?<secret>.*)", RegexOptions.Compiled, TimeSpan.FromSeconds(5));
+        private readonly ILogger logger;
+        private readonly ISecretClientProvider secretClientProvider;
+
+        public PostConfigureKeyVaultSettings(ILogger<PostConfigureKeyVaultSettings<T>> logger, ISecretClientProvider secretClientProvider)
+        {
+            this.logger = logger;
+            this.secretClientProvider = secretClientProvider;
+        }
+
+        public void PostConfigure(string name, T options)
+        {
+            var stringProperties = typeof(T)
+                .GetProperties()
+                .Where(x => x.PropertyType == typeof(string));
+
+            foreach (var property in stringProperties)
+            {
+                var value = (string)property.GetValue(options);
+
+                if (value != null)
+                {
+                    var match = secretRegex.Match(value);
+
+                    if (match.Success)
+                    {
+                        var vaultUrl = match.Groups["vault"].Value;
+                        var secretName = match.Groups["secret"].Value;
+
+                        try
+                        {
+                            var secretClient = this.secretClientProvider.GetSecretClient(new Uri(vaultUrl));
+                            this.logger.LogInformation("Replacing setting property {PropertyName} with value from secret {SecretUrl}", property.Name, value);
+
+                            var response = secretClient.GetSecret(secretName);
```

Comment:
At the risk of premature optimization, is it worth doing this async? Startup time could get high if there are a lot of secrets + client instantiations.

Reply:
OptionsManager<T> caches each named instance of IOptions<T>

https://github.com/dotnet/runtime/blob/main/src/libraries/Microsoft.Extensions.Options/src/OptionsManager.cs

I think this will prevent PostConfig from running multiple times